src/librustc/ty/context.rs

```diff
@@ -344,6 +345,9 @@ pub struct TypeckTables<'tcx> {
     /// method calls, including those of overloaded operators.
     type_dependent_defs: ItemLocalMap<Def>,
 
+    /// Stores the canonicalized types provided by the user.
+    user_provided_tys: ItemLocalMap<CanonicalTy<'tcx>>,
```

Comment:
Maybe add a "see also: the `UserAssertTy` statement in MIR" or something like that?